### PR TITLE
bpo-45919: Use WinAPI GetFileType() in is_valid_fd()

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue45919

### CODE REVIEW

**Comment by corona10** [comment]
I got the same result by measuring QueryPerformanceCounter.
5-6x+ faster

File: Python/pylifecycle.c [hunk 1]
```diff
@@ -2199,6 +2198,13 @@ is_valid_fd(int fd)
     _Py_END_SUPPRESS_IPH
 
     return (fd2 >= 0);
+#elif defined(MS_WINDOWS)
+    HANDLE hfile;
+    _Py_BEGIN_SUPPRESS_IPH
+    hfile = (HANDLE)_get_osfhandle(fd);
+    _Py_END_SUPPRESS_IPH
+    return (hfile != INVALID_HANDLE_VALUE &&
+              GetFileType(hfile) != FILE_TYPE_UNKNOWN);
```
Comment (vstinner) [inline Python/pylifecycle.c]: I have a personal preference for:

```suggestion
    return (hfile != INVALID_HANDLE_VALUE
            && GetFileType(hfile) != FILE_TYPE_UNKNOWN);
```

**Review by vstinner — APPROVED** [review]
LGTM. I just proposed a minor coding style change (it's up to you).

**Comment by eryksun** [comment]
The `_Py_BEGIN_SUPPRESS_IPH` and `_Py_END_SUPPRESS_IPH` macros are used to temporarily disable the MSVC invalid parameter handler for the `_get_osfhandle()` call. The use of these macros can be removed from the cases that are limited to POSIX systems.

**Comment by eryksun** [comment]
The comment that discusses the function should be updated to remove Windows from the platforms that use `dup()`.

**Comment by eryksun** [comment]
Was it verified that `fcntl(fd, F_GETFD)` addresses the case that forced macOS to use `fstat()` instead of `dup()`? If not, `create_stdio()` could fail with an `EBADF` error in macOS.

**Comment by vstinner** [comment]
> Was it verified that fcntl(fd, F_GETFD) addresses the case that forced macOS to use fstat() instead of dup()? If not, create_stdio() could fail with an EBADF error in macOS.

See the comment:

```
   bpo-30225: On macOS Tiger, when stdout is redirected to a pipe and the other
   side of the pipe is closed, dup(1) succeed, whereas fstat(1, &st) fails with
   EBADF. FreeBSD has similar issue (bpo-32849).
```

In case of doubt, I prefer to continue using fcntl() which always works. On BSD systems, dup() seems to implement less checks.

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by eryksun** [comment]
> In case of doubt, I prefer to continue using fcntl() which always works. On BSD systems, dup() seems to implement less checks.

@vstinner, the original change was to use `fstat()` instead of `dup()`. bpo-30225 makes no reference to `fcntl()`. @tiran added the `fcntl()` call in bpo-45915, with `__APPLE__` in the allowed platforms, but there was no discussion that it actually checks the pipe state in macOS to handle the case of bpo-30225.

**Comment by eryksun** [comment]
> The comment that discusses the function should be updated to remove Windows from the platforms that use `dup()`.

The out of date comment was not updated in the merge.

**Comment by vstinner** [comment]
> The out of date comment was not updated in the merge.

@corona10 wrote https://github.com/python/cpython/pull/30090

**Pull request merged**